Remove direct access to static config in code

mansu opened the pull request:
Remove direct access to `KaldbConfig.get()` in the code since it introduces tight coupling between classes and makes it harder to test code. Instead, pass the protobuf object around instead.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/slackhq/kaldb?pullRequest=242) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/slackhq/kaldb?pullRequest=242) before we can accept your contribution.<br/><hr/>**Suman Karumuri** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/slackhq/kaldb?pullRequest=242) it.</sub>

vthacker commented:
I think we need to fix `Kaldb#start` method - this still uses a static reference `KaldbConfig.get()` - this is important to remove since we want to be able to finally reuse `Kaldb.java` in test cases rather than create bespoke servers for testing that have different threading models than production.

vthacker commented:
What's to stop us from mistakenly calling `KaldbConfig#get` and accessing the value statically ? Shouldn't `_instance` not be a static anymore?

vthacker commented:
Looks like CLA assistant is unhappy as well

bryanlb commented on kaldb/src/main/java/com/slack/kaldb/chunkManager/CachingChunkManager.java:
```diff
@@ -95,15 +94,17 @@ protected void shutDown() throws Exception {
-      KaldbConfigs.ServerConfig serverConfig)
+      KaldbConfigs.ServerConfig serverConfig,
+      KaldbConfigs.S3Config s3Config,
+      KaldbConfigs.CacheConfig cacheConfig)
```
Why are we passing both `serverConfig` and `cacheConfig`, as `serverConfig` is a property of `cacheConfig`?

mansu replied:
Good point. Removed serverConfig param.

bryanlb commented on kaldb/src/main/java/com/slack/kaldb/server/KaldbIndexer.java:
```diff
@@ -50,8 +50,9 @@
-    String dataTransformerConfig = KaldbConfig.get().getIndexerConfig().getDataTransformer();
+  public static LogMessageTransformer getLogMessageTransformer(
+      KaldbConfigs.IndexerConfig indexerConfig) {
+    String dataTransformerConfig = indexerConfig.getDataTransformer();
```
Why are we passing the `indexerConfig` if we only use the `dataTransformer` string? This should only require the string config to be passed.

mansu replied:
Passing in most general type for now since there may be more params in future,

bryanlb replied:
As above, let's pass in the most specific argument. This case especially since this code will be moving out of the indexer entirely.

mansu replied:
Done.

bryanlb commented on kaldb/src/test/java/com/slack/kaldb/testlib/KaldbConfigUtil.java:
```diff
@@ -88,4 +70,26 @@ public static void initEmptyIndexerConfig() throws InvalidProtocolBufferExceptio
+    return makeIndexerConfig(TEST_INDEXER_PORT);
+  }
+
+  public static KaldbConfigs.IndexerConfig makeIndexerConfig(int indexerPort) {
```
It would be better if we called these `get...Config`, as that more clearly indicates they do not mutate state. `Make` is significantly more ambiguous, and requires reviewing the source to determine if it does indeed deal with state or is static.

mansu replied:
In general, get is used for a field access and using make for static methods. Also, the name is more consistent with existing convention, so left it as is.

bryanlb replied:
This is somewhat of an anti-pattern, as `get` is expected to be used for methods that return a "non-boolean function or attribute of the object on which they're invoked." Using `make` is fine, but `get` would be a more appropriate use.

bryanlb commented on kaldb/src/main/java/com/slack/kaldb/chunkManager/CachingChunkManager.java:
```diff
@@ -95,15 +94,17 @@ protected void shutDown() throws Exception {
-        KaldbConfig.get().getS3Config().getS3Bucket(),
-        KaldbConfig.get().getCacheConfig().getDataDirectory(),
-        KaldbConfig.get().getCacheConfig().getSlotsPerInstance());
+        s3Config.getS3Bucket(),
```
If we only need the `s3Config`, the `dataDirectory`, and `slotsPerInstance` can we just pass those instead?

mansu replied:
Yeah, I considered it and then decided to pass most general object, so it's more future proof when we add more params to this object.

bryanlb replied:
Let's pass the most specific object possible for all of these methods. We do not expect to be changing the behavior of these classes significantly as they are functionally complete at this time. 

It is not a significant amount of work to pass in an additional argument if needed in the future, and it will help when writing and troubleshooting if only the required arguments are passed.

bryanlb replied:
This one we agreed to leave for now, and revisit later with potentially a better scoped local config argument.

bryanlb commented on kaldb/src/main/java/com/slack/kaldb/logstore/LuceneIndexStoreImpl.java:
```diff
@@ -59,12 +60,13 @@
 
-  public static LuceneIndexStoreImpl makeLogStore(File dataDirectory, MeterRegistry metricsRegistry)
+  public static LuceneIndexStoreImpl makeLogStore(
+      File dataDirectory, KaldbConfigs.IndexerConfig indexerConfig, MeterRegistry metricsRegistry)
```
This should just take the commit duration and refresh duration if those are the only arguments we need.

mansu replied:
Passing in most general type here in case we add more params in future. Ideally, we should refactor these params into it's own Lucene config.

mansu replied:
done.

bryanlb replied:
It looks like this wasn't entirely updated, we didn't update the method to take the newly created `LuceneConfig`?

mansu commented:
> I think there are 4 references to `KaldbConfig.initFromConfigObject` in the tests that we need to remove - Those calls were made to workaround the problem that we had one static config. I'd be curious to know if the tests still pass once this is removed.

Removed the calls to `initFromConfigObject` from everywhere and made it a private method.

>I think we need to fix Kaldb#start method - this still uses a static reference KaldbConfig.get() - this is important to remove since we want to be able to finally reuse Kaldb.java in test cases rather than create bespoke servers for testing that have different threading models than production.

Good point. Removed. 


>What's to stop us from mistakenly calling KaldbConfig#get and accessing the value statically ? Shouldn't _instance not be a static anymore?
Moved KaldbConfig into server config and made this method package private. So, no one outside the package should be able to call KaldbConfig.get from the rest of the code.

vthacker commented on kaldb/src/main/java/com/slack/kaldb/chunkManager/ChunkManager.java:
```diff
@@ -1,6 +1,6 @@
 package com.slack.kaldb.chunkManager;
 
-import static com.slack.kaldb.config.KaldbConfig.LOCAL_QUERY_TIMEOUT_DURATION;
+import static com.slack.kaldb.server.KaldbConfig.LOCAL_QUERY_TIMEOUT_DURATION;
```
What's the reason here to change the package of `KaldbConfigs` class?  It seems better suited for the `config` package right?

vthacker replied:
but we literally are using it outside the package :) I don't get the reasoning . Could you please explain more about the change motivation

mansu replied:
These are static fields. The other methods are package private.

vthacker commented on kaldb/src/main/java/com/slack/kaldb/server/Kaldb.java:
```diff
@@ -75,7 +75,8 @@ public static void main(String[] args) throws Exception {
-    Kaldb kalDb = new Kaldb(configFilePath);
+    KaldbConfig.initFromFile(configFilePath);
+    Kaldb kalDb = new Kaldb(KaldbConfig.get());
     kalDb.start();
```
I think we want `start` to take `KaldbConfig` as an argument so that we can potentially reuse the `start` method in the test cases

mansu replied:
Sure, we can make that change when we change the tests. For now the tests work well.

bryanlb reviewed (approved):
🐑
